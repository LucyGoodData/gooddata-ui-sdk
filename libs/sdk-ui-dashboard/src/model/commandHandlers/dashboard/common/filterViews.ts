// (C) 2024 GoodData Corporation

import {
    IDashboard,
    IDashboardFilterView,
    isFilterContext,
    ISettings,
    isDashboardAttributeFilter,
    IDashboardAttributeFilter,
    FilterContextItem,
    IDashboardDateFilter,
    isDashboardDateFilter,
    areObjRefsEqual,
    IFilterContext,
} from "@gooddata/sdk-model";
import omit from "lodash/omit.js";
import isEqual from "lodash/isEqual.js";

const findMatchingAttributeFilterByLocalIdentifier = (
    filter: IDashboardAttributeFilter,
    viewFilters: FilterContextItem[],
): IDashboardAttributeFilter | undefined =>
    viewFilters.find(
        (item) =>
            isDashboardAttributeFilter(item) &&
            item.attributeFilter.localIdentifier === filter.attributeFilter.localIdentifier,
    ) as IDashboardAttributeFilter;

// date filters do not have localIdentifier set, compare them by dataSet instead
const findMatchingDateFilterByDataSet = (
    filter: IDashboardDateFilter,
    viewFilters: FilterContextItem[],
): IDashboardDateFilter | undefined =>
    viewFilters.find(
        (item) =>
            isDashboardDateFilter(item) &&
            ((item.dateFilter.dataSet === undefined && filter.dateFilter.dataSet === undefined) ||
                areObjRefsEqual(item.dateFilter.dataSet, filter.dateFilter.dataSet)),
    ) as IDashboardDateFilter;

const OMITTED_ATTRIBUTE_FILTER_PATHS = [
    "attributeFilter.attributeElements",
    "attributeFilter.negativeSelection",
];

const hasSameAttributeFilterConfiguration = (
    filterA: IDashboardAttributeFilter,
    filterB?: IDashboardAttributeFilter,
) => {
    return isEqual(
        omit(filterA, OMITTED_ATTRIBUTE_FILTER_PATHS),
        omit(filterB, OMITTED_ATTRIBUTE_FILTER_PATHS),
    );
};

export const changeFilterContextSelection = (
    filterContext: IFilterContext,
    filterViewFilters: FilterContextItem[],
): IFilterContext => {
    return {
        ...filterContext,
        filters: filterContext.filters.map((filter) => {
            if (isDashboardAttributeFilter(filter)) {
                const viewFilter = findMatchingAttributeFilterByLocalIdentifier(filter, filterViewFilters);
                if (viewFilter !== undefined && hasSameAttributeFilterConfiguration(filter, viewFilter)) {
                    return {
                        attributeFilter: {
                            ...filter.attributeFilter,
                            attributeElements: viewFilter.attributeFilter.attributeElements,
                            negativeSelection: viewFilter.attributeFilter.negativeSelection,
                        },
                    };
                }
            } else {
                const viewFilter = findMatchingDateFilterByDataSet(filter, filterViewFilters);
                if (viewFilter) {
                    return {
                        dateFilter: {
                            ...filter.dateFilter,
                            from: viewFilter.dateFilter.from,
                            to: viewFilter.dateFilter.to,
                            granularity: viewFilter.dateFilter.granularity,
                            type: viewFilter.dateFilter.type,
                        },
                    };
                }
            }
            return filter;
        }),
    };
};

export function applyDefaultFilterView(
    dashboard: IDashboard,
    filterViews: IDashboardFilterView[],
    settings: ISettings,
): IDashboard {
    // find first default filter view (in case metadata are not consistent and there are more than one)
    const defaultFilterView = filterViews.find((view) => view.isDefault);
    const areFilterViewsEnabled = settings.enableDashboardFilterViews;

    return areFilterViewsEnabled && defaultFilterView && isFilterContext(dashboard.filterContext)
        ? {
              ...dashboard,
              filterContext: changeFilterContextSelection(
                  dashboard.filterContext,
                  defaultFilterView.filterContext.filters,
              ),
          }
        : dashboard;
}
