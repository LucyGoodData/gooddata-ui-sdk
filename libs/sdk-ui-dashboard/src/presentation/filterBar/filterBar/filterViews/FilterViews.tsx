// (C) 2024 GoodData Corporation

import React, { useCallback } from "react";
import { FormattedMessage } from "react-intl";
import cx from "classnames";
import { IAlignPoint, DropdownButton, useMediaQuery } from "@gooddata/sdk-ui-kit";
import { IDashboardFilterView } from "@gooddata/sdk-model";

import { ConfigurationBubble } from "../../../widget/common/configuration/ConfigurationBubble.js";
import {
    useDashboardSelector,
    selectFilterViews,
    selectIsInEditMode,
    useDashboardDispatch,
    uiActions,
    selectDisableFilterViews,
    selectIsFilterViewsDialogOpen,
    selectFilterViewsDialogMode,
    selectIsNewDashboard,
} from "../../../../model/index.js";

import { FilterViewsList } from "./FilterViewsList.js";
import { useFilterViewsToastMessages } from "./useFilterViewsToastMessages.js";
import { AddFilterView } from "./AddFilterView.js";

const BUBBLE_ALIGN_POINTS: IAlignPoint[] = [{ align: "br tr", offset: { x: -27, y: -10 } }];

const DropdownButtonLabel: React.FC<{ filterViews: IDashboardFilterView[] }> = ({ filterViews }) => {
    return filterViews.length === 0 ? (
        <FormattedMessage id="filters.filterViews.dropdown.buttonEmpty" />
    ) : (
        <FormattedMessage
            id="filters.filterViews.dropdown.button"
            values={{ count: filterViews?.length ?? 0 }}
        />
    );
};

const useCallbacks = () => {
    const dispatch = useDashboardDispatch();

    const toggleDialog = useCallback(() => dispatch(uiActions.toggleFilterViewsDialog()), [dispatch]);

    const openListDialog = useCallback(
        () =>
            dispatch(
                uiActions.toggleFilterViewsDialog({
                    open: true,
                    mode: "list",
                }),
            ),
        [dispatch],
    );

    const openAddDialog = useCallback(
        () =>
            dispatch(
                uiActions.toggleFilterViewsDialog({
                    open: true,
                    mode: "add",
                }),
            ),
        [dispatch],
    );

    const closeDialog = useCallback(
        () =>
            dispatch(
                uiActions.toggleFilterViewsDialog({
                    open: false,
                }),
            ),
        [dispatch],
    );

    return {
        toggleDialog,
        openListDialog,
        openAddDialog,
        closeDialog,
    };
};

export const FilterViews: React.FC = () => {
    const isDialogOpen = useDashboardSelector(selectIsFilterViewsDialogOpen);
    const dialogMode = useDashboardSelector(selectFilterViewsDialogMode);
    const isDashboardEditMode = useDashboardSelector(selectIsInEditMode);
    const filterViews = useDashboardSelector(selectFilterViews);
    const isFilterViewsEnabledForDashboard = !useDashboardSelector(selectDisableFilterViews);
    const isNewDashboard = useDashboardSelector(selectIsNewDashboard);
    const isMobile = useMediaQuery("mobileDevice");
    const { toggleDialog, openAddDialog, openListDialog, closeDialog } = useCallbacks();

    useFilterViewsToastMessages();

    // On mobile the dropdown button is not on filter bar but rendered inside the "..." dashboard menu.
    // The dropdown body must still be rendered so the body can be opened when user uses the menu.
    // Also, the menu is not available for new dashboard as new filter view can be saved only for an existing
    // dashboard with ref id.
    const showDropdownButton = isFilterViewsEnabledForDashboard && !isNewDashboard && !isMobile;

    const buttonClassNames = cx(
        "gd-filter-views-button",
        "gd-filter-views__dropdown-anchor",
        "gd-button-large",
        {
            "gd-filter-views-button--open": isDialogOpen,
        },
    );

    return (
        <div className="gd-filter-views">
            {showDropdownButton ? (
                <>
                    <DropdownButton onClick={toggleDialog} className={buttonClassNames} isOpen={isDialogOpen}>
                        <DropdownButtonLabel filterViews={filterViews} />
                    </DropdownButton>
                    {isDashboardEditMode ? <div className="gd-filters-views__panel__divider" /> : null}
                </>
            ) : (
                <div className="gd-filter-views__dropdown-anchor" />
            )}
            {isDialogOpen ? (
                <ConfigurationBubble
                    classNames="gd-filters-views__panel"
                    onClose={closeDialog}
                    alignTo=".gd-filter-views__dropdown-anchor"
                    alignPoints={BUBBLE_ALIGN_POINTS}
                >
                    {dialogMode === "add" ? (
                        <AddFilterView onClose={openListDialog} />
                    ) : (
                        <FilterViewsList
                            filterViews={filterViews}
                            onAddNew={openAddDialog}
                            onClose={closeDialog}
                        />
                    )}
                </ConfigurationBubble>
            ) : null}
        </div>
    );
};
