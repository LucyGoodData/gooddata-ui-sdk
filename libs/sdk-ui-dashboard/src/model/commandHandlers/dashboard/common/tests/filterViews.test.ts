// (C) 2024 GoodData Corporation

import { it, describe, expect } from "vitest";
import { changeFilterContextSelection } from "../filterViews";
import { FilterContextItem } from "@gooddata/sdk-model";
import { filterContext } from "./filterViews.mock";

describe("filterViews", () => {
    describe("changeFilterContextSelection", () => {
        it("should set selection of matching filters in provided filters context and leave the rest of its filters as is", () => {
            const filters: FilterContextItem[] = [
                {
                    dateFilter: {
                        granularity: "GDC.time.year",
                        type: "relative",
                        from: -1,
                        to: -1,
                    },
                },
                {
                    attributeFilter: {
                        attributeElements: { uris: ["Advertising"] },
                        displayForm: { identifier: "campaign_channels.category", type: "displayForm" },
                        negativeSelection: true,
                        localIdentifier: "86e90c9b864e4701affced5e55b36b9c",
                        // the filter setting no longer matches the one in the filter context, therefore will not be applied
                        selectionMode: "multi",
                    },
                },
                {
                    attributeFilter: {
                        attributeElements: { uris: ["Midwest"] },
                        displayForm: { identifier: "region", type: "displayForm" },
                        negativeSelection: false,
                        localIdentifier: "a01fc28d9a8244b99179d3f7320b3400",
                        selectionMode: "multi",
                    },
                },
                {
                    dateFilter: {
                        dataSet: { identifier: "date", type: "dataSet" },
                        type: "relative",
                        granularity: "GDC.time.year",
                        from: -1,
                        to: -1,
                    },
                },
                {
                    attributeFilter: {
                        attributeElements: { uris: ["C"] },
                        displayForm: { identifier: "negative-filter", type: "displayForm" },
                        negativeSelection: false,
                        localIdentifier: "b01fc28d9a8244b99179d3f7320b3400",
                        selectionMode: "multi",
                    },
                },
            ];
            expect(changeFilterContextSelection(filterContext, filters)).toMatchSnapshot();
        });
    });
});
