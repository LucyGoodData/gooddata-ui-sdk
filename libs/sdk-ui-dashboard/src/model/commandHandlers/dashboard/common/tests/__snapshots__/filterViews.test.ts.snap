// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`filterViews > changeFilterContextSelection > should set selection of matching filters in provided filters context and leave the rest of its filters as is 1`] = `
{
  "description": "",
  "filters": [
    {
      "dateFilter": {
        "from": -1,
        "granularity": "GDC.time.year",
        "to": -1,
        "type": "relative",
      },
    },
    {
      "attributeFilter": {
        "attributeElements": {
          "uris": [
            "Content",
          ],
        },
        "displayForm": {
          "identifier": "campaign_channels.category",
          "type": "displayForm",
        },
        "localIdentifier": "86e90c9b864e4701affced5e55b36b9c",
        "negativeSelection": false,
        "selectionMode": "single",
      },
    },
    {
      "attributeFilter": {
        "attributeElements": {
          "uris": [
            "Midwest",
          ],
        },
        "displayForm": {
          "identifier": "region",
          "type": "displayForm",
        },
        "localIdentifier": "a01fc28d9a8244b99179d3f7320b3400",
        "negativeSelection": false,
        "selectionMode": "multi",
      },
    },
    {
      "attributeFilter": {
        "attributeElements": {
          "uris": [
            "C",
          ],
        },
        "displayForm": {
          "identifier": "negative-filter",
          "type": "displayForm",
        },
        "localIdentifier": "b01fc28d9a8244b99179d3f7320b3400",
        "negativeSelection": false,
        "selectionMode": "multi",
      },
    },
    {
      "dateFilter": {
        "dataSet": {
          "identifier": "date",
          "type": "dataSet",
        },
        "from": -1,
        "granularity": "GDC.time.year",
        "to": -1,
        "type": "relative",
      },
    },
    {
      "dateFilter": {
        "dataSet": {
          "identifier": "ignored",
          "type": "dataSet",
        },
        "from": "2023-06-10",
        "granularity": "GDC.time.date",
        "to": "2023-02-11",
        "type": "absolute",
      },
    },
    {
      "attributeFilter": {
        "attributeElements": {
          "uris": [],
        },
        "displayForm": {
          "identifier": "order_id",
          "type": "displayForm",
        },
        "localIdentifier": "87fbd090f82a4425822ecf19f4f9e123",
        "negativeSelection": true,
        "selectionMode": "multi",
      },
    },
  ],
  "identifier": "1c07035a-d48c-48f9-97ac-ca6a146a1b17",
  "ref": {
    "identifier": "1c07035a-d48c-48f9-97ac-ca6a146a1b17",
    "type": "filterContext",
  },
  "title": "filterContext",
  "uri": "https://internal-testing.staging.stg11.panther.intgdc.com/api/v1/entities/workspaces/130ee9a3d69b4508aed431773a660706/filterContexts/1c07035a-d48c-48f9-97ac-ca6a146a1b17",
}
`;
